### ISSUE TITLE
perf(pcs): verify path runs eagerly — composites interpret their decomposition in Python

### ISSUE THREAD

**Issue by chokobole** [fractalyze/zorch#140]
### Context

With the composite jaxlib (zkx#517 lineage) the jagged opening suite's
compile cost is solved — a cold `test_round_trip_verifies` is 265 s of which
only ~21 s is compile. The remaining ~244 s is execution, and a py-spy native
profile of the warm run attributes it almost entirely to **Python-side eager
dispatch**, not compiled code:

- `verify` subtree: **45%** of samples; `open`: **27%**. Actual compiled
  execution (`_pjit_call_impl` and below): **~1–3%**. Runtime compile: ~0.1%.
- The hot chain is `bind → process_primitive → _composite_impl →
  jaxpr_as_fun → eval_jaxpr`: an **eager `lax.composite` interprets its
  decomposition jaxpr op-by-op in Python**. Outside a `@jit` zone the
  fusion-by-construction markers not only miss the fused emitters — they
  make things slower than plain ops.
- `BasefoldVerifier.verify`'s `roots_ok` re-traces per call: **10%** of all
  samples sit under `vmap` trace machinery (`_batch_outer`,
  `flatten_fun_for_vmap`), because `jax.vmap(self.tree.reconstruct_root)` is
  constructed inside the call (fresh function identity → no trace cache).
  `BasefoldProver.open`'s per-fold `jax.vmap(lambda i: ...)` has the same
  shape.

### Spec

The prover deliberately uses per-round `@jit` zones plus a Python loop (see
`zorch/pcs/jagged/prover.py` round-loop comments) — that part is fine. The
verify path has no `@jit` at all:

- [x] Put `JaggedPcsVerifier.verify` / `stacked_verify` /
  `BasefoldVerifier.verify` under `@jit` zones (whole-verify or per-round,
  matching the prover's structure).
- [x] Hoist vmapped callables (`reconstruct_root`, the per-fold open lambda)
  to construction time so their traces cache.
- [x] Re-measure: warm `test_round_trip_verifies` execution, and the suite.
- [x] Extend the jit zones to fri (`FriVerifier.verify`,
  `FriProver.commit`/`open` — untouched by `perf/jit-verify-path`; CPU
  profile: verifier alone is 122 s of a 170 s e2e test).
- [ ] Re-measure the CPU suites (`JAX_PLATFORMS=cpu`): fri single e2e
  170 s → target ≤ ~30 s; basefold+jagged 40 tests / 4704 s cold → minutes.

### Expected impact

The compiled work itself is tiny (the whole graph compiles in ~21 s and the
fused poseidon2 kernels are fast); the 4-minute round trip should drop by
roughly the eager fraction (>90%) once verify executes inside jit.

### Relevant knowledge

- Eager-jit slowness has bitten before in riscv-witness ("not hang, just slow
  eager jit"). Same signature: time in `core.bind`/tracing, none in
  executables.

**Comment by chokobole** [comment]
### Session 2026-06-05 15:00 progress

- Done: verify-path jit zones + construction-time vmap hoists on branch `perf/jit-verify-path` (460fd9d). Correctness: basefold/fri/stacked suites 15/15 on the composite GPU plugin; a CheapPermutation jagged round trip (log_s 5 and 3) verifies, a tampered `dense_eval` rejects, and repeated verifies hold exactly one trace (the re-trace pathology is gone).
- Next session: re-measure warm `test_round_trip_verifies` + the opening suite on GPU once the zkx#511 fix lands, then `/create-pr`.
- Blocker: fractalyze/zkx#511 — fresh-compiling any EF sponge-hash fusion segfaults the GPU plugin (`Poseidon2SpongeHashFusion::EmitEntryFunction`, GPU twin documented on that issue). `prover.commit` dies at compile on baseline `main` too, so the opening suite is unrunnable on the current toolchain regardless of this branch; earlier runs were masked by warm cache entries.

**Comment by chokobole** [comment]
### CPU e2e suites hit the same wall (fri profile, 2026-06-05)

cProfile of one fri e2e test on CPU (`JAX_PLATFORMS=cpu`, opt toolchain, jax @ da27584) — 170 s total:

| where | time |
|---|---|
| XLA compile (`backend_compile_and_load`) | **4.5 s** — 244 compiles, ~19 ms each |
| Python eager dispatch | **~160 s** — 1.47M `core.bind`, 558K `process_primitive` |
| └ `poseidon2.permute` ×21 via `fused_region` | 152 s (~7 s/call) |
| └ verifier `merkle.reconstruct_root` ×4 | 122 s of 169 s |

Same mechanism as the GPU finding above; CPU-specific consequences:

- **The persistent compilation cache cannot help these suites**: no single
  compile reaches the 1 s `jax_persistent_cache_min_compile_time_secs`
  threshold — 0 entries banked across a full 7-test fri run with
  `JAX_COMPILATION_CACHE_DIR` set.
- The basefold+jagged CPU baseline is 40 tests / 4704 s cold (#143 wrap);
  `perf/jit-verify-path` should recover most of it — worth a CPU re-measure
  alongside the GPU one.
- **fri is not covered by the branch**: `pcs/fri/verifier.py` is untouched —
  `FriVerifier.verify` has no jit zone and builds
  `jax.vmap(params.tree.reconstruct_root)` inside the call (the same
  fresh-identity re-trace shape named above). `FriProver.commit`/`open` are
  eager too, with a per-call `jax.vmap(lambda i: ...)` in `open_batch`.

Extended the Spec checklist accordingly. Repro:

```sh
JAX_PLATFORMS=cpu python -m cProfile -o /tmp/t.prof \
  -m pytest "zorch/pcs/fri/testing/fri_test.py::FriRoundTripTest::test_honest_opening_verifies" -q
python -c "import pstats; s=pstats.Stats('/tmp/t.prof'); s.sort_stats('tottime'); s.print_stats(15)"
```

Projected for fri: 170 s/test → ~15–25 s once prove+verify run under jit.
Data origin: #143 wrap comment (2026-06-05, "filing separately" → here).

**Comment by chokobole** [comment]
### GPU re-measure (2026-06-06, zkx main 40b85737 = #514+#520+#522+#523, RTX 5090)

| metric | before (eager verify, #142-era toolchain) | after (`perf/jit-verify-path` 460fd9d, EF fused) |
|---|---|---|
| warm `test_round_trip_verifies` | ~244 s | **123.6 s** |
| warm suite (7 tests) | 1011 s | **857.8 s** |
| cold suite | unrunnable (zkx#511 segfault on fresh compile) | 1169.9 s |

- The eager-dispatch pathology is gone: each verify zone traces once (jit cache size 1 across repeated verifies) and replays compiled; cold compile no longer hits the unrolled-inline cliff now that EF poseidon2 state takes the fused emitters on both backends (zkx#517).
- The issue's >90% projection did not materialize because the py-spy CPU profile could not see device-side time: with Python dispatch removed, the residual ~120 s/verify is GPU **launch-bound execution** — the verify graph lowers to many small sequential kernels (transcript scan steps, per-query/layer Merkle reconstruct chains). Confirmed across 5 warm verify-only tests at 118–132 s each.
- That residual is the next lever (fusion-by-construction for the verify replay, cf. the #523 register-resident pattern) — needs its own issue once the fri checkboxes here are done.

**Comment by chokobole** [comment]
### fri extension + CPU re-measure (2026-06-06)

fri got the same treatment (9e614ff): `FriProof` registered as a pytree, the per-call vmap closures hoisted, and `commit`/`open`/`_verify_one` jitted per-poly (fri owns its zones — unlike basefold, nothing encloses it).

| metric (CPU, dist311-zkx487 wheel) | before | after |
|---|---|---|
| fri e2e (`test_honest_opening_verifies`), warm | 170 s | **57.9 s** |
| fri suite (4 tests), warm | — | 192 s |

GPU: fri 4/4 (192 s incl. cold compile of the new zones on zkx main).

The ≤ ~30 s target and the basefold+jagged CPU suite re-measure are gated on a post-zkx#520 CPU wheel (the dist311-zkx487 wheel predates the fused-permute routing, so the residual is unfused-permute execution, not dispatch) — carried to #163 alongside the GPU launch-bound verify work, which is the remaining lever on both backends.
